### ISSUE TITLE
Fix prod double backward when there are 2+ zeros

### ISSUE THREAD

**Issue by zou3519** [pytorch/pytorch#106789]
### 🐛 Describe the bug

When there are two or more zeros, prod's gradient does not require grad. This is equivalent to saying that the double backwards is 0, i.e. there are no gradients to be computed. However, this is wrong; the double backwards should not be zero.
```
import torch

x = torch.tensor([2., 3, 0, 0], requires_grad=True)
y = torch.prod(x)
gx, = torch.autograd.grad(y.sum(), x, create_graph=True)
```

### Versions

main

**Comment by lezcano** [comment]
Fixed in https://github.com/pytorch/pytorch/issues/106789
